### PR TITLE
Add infrastructure and some documentation

### PR DESCRIPTION
**Pull request by sethrj**

### CODE REVIEW

**Comment by sethrj** [comment]
OK @tmdelellis I fixed the code and squashed the commits. This is ready for for-real reviews now :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+foreach(_LANG CUDA CXX)
+  set(CMAKE_${_LANG}_STANDARD 14)
+  set(CMAKE_${_LANG}_EXTENSIONS OFF)
+endforeach()
+
+#---------------------------------------------------------------------------##
+# DEPENDENCIES
+#---------------------------------------------------------------------------##
+
+if(CELERITAS_USE_CUDA)
```
[the remark this answers is not in the document]
Comment (tmdelellis) [reply CMakeLists.txt:58]: One issue we may encounter downstream is that when the compiler is `mpic++`, the link step for tests will be done with the `nvcc` host compiler (will probably default to `/usr/bin/g++` or `/usr/bin/c++`), and we will get link errors.  Unfortunately, I haven't found a way to make CMake handle this automatically, `check_language` sets the CUDA host compiler, but it won't make it the same as the regular compiler.

One additional thing to note, if you are using clang, you will now get this error:
```
-- Check for working CUDA compiler: /usr/local/cuda-10.2/bin/nvcc -- broken
CMake Error at /projects/spack/opt/spack/gcc-8.3.1/cmake/kh6zyzk/share/cmake-3.16/Modules/CMakeTestCUDACompiler.cmake:46 (message):
  The CUDA compiler

    "/usr/local/cuda-10.2/bin/nvcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/9te/celeritas/_build/CMakeFiles/CMakeTmp

    Run Build Command(s):/projects/spack/opt/spack/gcc-8.3.1/ninja-fortran/5vubfit/bin/ninja-build cmTC_e947d && [1/2] Building CUDA object CMakeFiles/cmTC_e947d.dir/main.cu.o
    FAILED: CMakeFiles/cmTC_e947d.dir/main.cu.o
    /usr/local/cuda-10.2/bin/nvcc -ccbin=/projects/spack/opt/spack/gcc-8.3.1/llvm/qqywtaq/bin/clang++   -arch=sm_35 -x cu -c main.cu -o CMakeFiles/cmTC_e947d.dir/main.cu.o && /usr/local/cuda-10.2/bin/nvcc -ccbin=/projects/spack/opt/spack/gcc-8.3.1/llvm/qqywtaq/bin/clang++   -arch=sm_35 -x cu -M main.cu -MT CMakeFiles/cmTC_e947d.dir/main.cu.o -o CMakeFiles/cmTC_e947d.dir/main.cu.o.d
    In file included from <built-in>:1:
    In file included from /usr/local/cuda-10.2/bin/../targets/x86_64-linux/include/cuda_runtime.h:83:
    /usr/local/cuda-10.2/bin/../targets/x86_64-linux/include/crt/host_config.h:146:2: error: -- unsupported clang version! clang version must be less than 9 and greater than 3.2
    #error -- unsupported clang version! clang version must be less than 9 and greater than 3.2
     ^
    1 error generated.
    ninja: build stopped: subcommand failed.
```
So we may want to install llvm 8 on emmet in your Spack distro if we want to use LLVM.  Using gcc works fine (I tested).

[the remark this answers is not in the document]
Comment (sethrj) [reply CMakeLists.txt:58]: By linking errors, do you mean with MPI? If the celeritas libraries are linked publicly against `MPI::C` (which they are) it should prevent that. If the compiled native NVCC code is ABI compatible with the host compiler, do we need to ensure the host compilers sync up? That seems like the sort of thing CMake should default to.

[the remark this answers is not in the document]
Comment (tmdelellis) [reply CMakeLists.txt:58]: Unfortunately, it doesn't. Also, when you compile with `mpic++` I've had errors that linking against `MPI::C` and `MPI::CXX` didn't fix.  But, nearer to home, if you compile with llvm right now as you have it, the Cuda random test will fail to link with the following error:
```
[32/32] Linking CUDA executable test/random_RadialDistribution_cuda
FAILED: test/random_RadialDistribution_cuda
: && /usr/bin/g++ -Wl,-rpath -Wl,/projects/spack/opt/spack/gcc-8.3.1/hwloc/ryw4s2i/lib -Wl,-rpath -Wl,/projects/spack/opt/spack/gcc-8.3.1/zlib/juqj6fv/lib -Wl,-rpath -Wl,/projects/spack/opt/spack/gcc-8.3.1/openmpi/aggdk4t/lib -L/projects/spack/opt/spack/gcc-8.3.1/hwloc/ryw4s2i/lib -L/projects/spack/opt/spack/gcc-8.3.1/zlib/juqj6fv/lib -pthread test/CMakeFiles/random_RadialDistribution_cuda.dir/random/RadialDistribution.test.cu.o -o test/random_RadialDistribution_cuda  -Wl,-rpath,/projects/spack/opt/spack/gcc-8.3.1/cmake/kh6zyzk/lib:/projects/spack/opt/spack/gcc-8.3.1/llvm/qqywtaq/lib:/projects/spack/opt/spack/gcc-8.3.1/vecgeom/iui3w3j/lib:/projects/spack/opt/spack/gcc-8.3.1/git-lfs/6rbzvtj/lib:/projects/spack/opt/spack/gcc-8.3.1/git/frxi5g2/lib:/usr/local/cuda-10.2/lib:/projects/spack/opt/spack/gcc-8.3.1/openmpi/aggdk4t/lib:/projects/spack/opt/spack/gcc-8.3.1/ninja-fortran/5vubfit/lib:/home/9te/celeritas/_build/test:/home/9te/celeritas/_build/src:/home/9te/celeritas/_build/lib  test/libCeleritasTest.so  src/libceleritas.so  /projects/spack/opt/spack/gcc-8.3.1/openmpi/aggdk4t/lib/libmpi_cxx.so  /projects/spack/opt/spack/gcc-8.3.1/openmpi/aggdk4t/lib/libmpi.so  lib/libgtestd.so  -lpthread -L"/usr/local/cuda-10.2/targets/x86_64-linux/lib/stubs" -L"/usr/local/cuda-10.2/targets/x86_64-linux/lib" -lcudadevrt -lcudart_static -lrt -lpthread -ldl && :
test/CMakeFiles/random_RadialDistribution_cuda.dir/random/RadialDistribution.test.cu.o: In function `testing::AssertionResult::AppendMessage(testing::Message const&)':
/home/9te/celeritas/celeritas/external/googletest/googletest/include/gtest/gtest.h:350: undefined reference to `testing::Message::GetString[abi:cxx11]() const'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
The problem is that its using `/usr/bin/g++` to link whereas `clang++` was used to build the other libraries and, for some reason, their symbols are not compatible.

[the remark this answers is not in the document]
Comment (sethrj) [reply CMakeLists.txt:58]: Ah I see, that makes sense: looks like they link to different implementations of the standard library.

[the remark this answers is not in the document]
Comment (sethrj) [reply CMakeLists.txt:56]: OK added and tested the host compiler defaulting to CXX_HOST! Also fixed a build failure when MPI was disabled.

**Pull request merged**